src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -964,6 +1085,64 @@ public DocumentId AddTextContainer(SourceTextContainer textContainer, string ful
                 return documentId;
             }
 
+            public void AddDynamicFile(IDynamicFileInfoProvider fileInfoProvider, DynamicFileInfo fileInfo, ImmutableArray<string> folders)
+            {
+                var documentInfo = CreateDocumentInfoFromFileInfo(fileInfo, folders.NullToEmpty());
+                var documentId = documentInfo.Id;
+
+                lock (_project._gate)
+                {
+                    var filePath = documentInfo.FilePath;
+                    if (_documentPathsToDocumentIds.ContainsKey(filePath))
+                    {
+                        throw new ArgumentException($"'{filePath}' has already been added to this project.", nameof(filePath));
+                    }
+
+                    _documentPathsToDocumentIds.Add(filePath, documentId);
+
+                    _documentIdToDynamicFileInfoProvider.Add(documentId, fileInfoProvider);
+
+                    if (_project._eventSubscriptionTracker.Add(fileInfoProvider))
+                    {
+                        // subscribe to the event when we use this provider the first time
+                        fileInfoProvider.Updated += _project.OnDynamicFileInfoUpdated;
```

Comment:
❔ Where is this event unregistered?

💭 It seems like this block could be extracted to a separate method. It seems strange to have event registration appear as a per-document action and unregistration appear as a per-project action. In reality it doesn't behave that way, but it's not immediately obvious.